Documentation for OQL expression syntax (#8520)
mendix/docs: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md

##### timezone

`timezone` specifies the time zone to use for the retrieval. This parameter is optional and defaults to the local time zone. It should be a string literal containing an IANA time zone. GMT offset time zones are not supported.

#### Examples

Assume the entity `Sales.Period` has 2 objects:

```sql
SELECT * FROM Sales.Period
```

| ID | Start               | End                 | Revenue |                                                        
|:---|---------------------|---------------------|---------|
| -  | 2024-05-02 00:00:00 | 2025-07-05 00:00:00 | 28      |
| -  | 2024-05-02 00:00:00 | 2024-06-02 15:12:45 | 10      |

We can use `DATEDIFF` to get the time interval between two dates:

```sql
SELECT DATEDIFF(MONTH , End, Start ) as difference FROM Sales.Period
```

| difference |                                                        
|------------|
| 14         |
| 1          |

This interval can be used to calculate the average revenue per month:

```sql
SELECT Revenue : DATEDIFF(MONTH, End, Start ) as avg_revenue FROM Sales.Period
```

| avg_revenue |                                                        
|-------------|
| 2           |

> | 10          |

Reviewer: Nice example.

{{% /alert %}}

### DATEPART

The `DATEPART` function retrieves a specified element from a date/time values. The return type is `INTEGER`.

#### Syntax

The syntax is as follows:

```sql
DATEPART ( datepart , date_expression [, timezone ] )
```

##### datepart

`datepart` specifies the part of the date/time value to retrieve. For possible values, see the [Example](#oql-datepart-example) below.

##### date_expression

`date_expression` specifies the date to retrieve an element from. The expression should resolve to a `DATETIME` value, string representations of `DATETIME` are accepted.

##### timezone

`timezone` specifies the time zone to use for the retrieval. This parameter is optional and defaults to the local time zone. It should be a string literal containing an IANA time zone. GMT offset time zones are not supported.

#### Examples{#oql-datepart-example}

| datepart     | Definition                                       | Example (Friday July 1, 2005, 16:34:20.356) |
|--------------|--------------------------------------------------|---------------------------------------------|

> | `YEAR`       |                                                  | 2005                                        |

Reviewer: Are there years where this won't work? e.g. 835 CE or 3500 BCE?

Reviewer 2: I think it really depends on the database, but I would rather not include such details

Reviewer: OK, let's leave it at this - we can't cover everything.

| `QUARTER`    | 1, 2, 3 or 4                                     | 3                                           |
| `MONTH`      | 1 to 12                                          | 7                                           |
| `DAYOFYEAR`  | 1 to 366                                         |                                             |
| `DAY`        | 1 to 31                                          | 5                                           |
| `WEEK`       | 1 to 53 (depends on the database implementation) |                                             |
| `WEEKDAY`    | 1 to 7 (1 = Sunday, 7 = Saturday)                | 6                                           |
| `HOUR`       | 0 to 23                                          | 16                                          |
| `MINUTE`     | 0 to 59                                          | 34                                          |
| `SECOND`     | 0 to 59                                          | 20                                          |

> | `MILISECOND` | 0 to 999                                         | 356                                         |

Reviewer: I corrected this to `MILLISECOND` above. Should it be
```suggestion
| `MILLISECOND` | 0 to 999                                         | 356                                         |
```
?

```sql
SELECT * FROM Sales.Reports
```

| ID | Text                          |                                                         
|----|-------------------------------|
| -  | "Performance is satisfactory" |
| -  | "Order has been completed"    |

Their length can be calculated with in an extra column

```sql
SELECT Text, LENGTH(Text) as text_length FROM Sales.Reports
```

| Text                          | text_length |                                                         
|-------------------------------|-------------|
| "Performance is satisfactory" | 27          |
| "Order has been completed"    | 24          |

### LOWER{#lower-function}

#### Description

Converts all uppercase characters in a given string to lowercase.

#### Syntax

The syntax is as follows:

```sql
LOWER ( expression )
```

`expression` specifies the string to convert.

#### Example

The function is useful to enforce consistent case for all strings, especially for comparisons. The query below would return no results in case-sensitive databases, as there is only a "Doe":

```sql
SELECT * FROM Sales.Customer WHERE LastName = 'doe'
```

Using `LOWER` this inconsistency can be fully avoided:

```sql

> SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'

Reviewer: Would it be a better practice to LOWER both sides of the = expression? (even through it is unnecessary in this example).
```suggestion
SELECT * FROM Sales.Customer WHERE LOWER(LastName) = LOWER('doe')
```

Reviewer 2: I'd say for understanding how LOWER works this example is good because if you have LOWER on both sides, who knows how it transforms the string

Reviewer: Good point. Let's leave it.

| 2025-07-05 00:00:00 | 28      |

### RANGEEND{#oql-rangeend}

Extracts the end value of a range parameter.

#### Syntax

```sql
RANGEEND ( $range )
```

`$range` specifies the range parameter.

#### Example

This query uses `$range` to retrieve all periods that end before the end value of `$range`:

```sql
SELECT End, Revenue FROM Sales.Period
WHERE End < RANGEEND($range)
```

| End                 | Revenue |
|---------------------|---------|
| 2024-06-02 15:12:45 | 10      |

This query uses `$range_past` to retrieve all periods that ended before the end date of `$range_past`:

```sql
SELECT End, Revenue FROM Sales.Period
WHERE End < RANGEEND($range_past)
```

| End                 | Revenue |
|---------------------|---------|
| 2024-06-02 15:12:45 | 10      |

### REPLACE

The REPLACE function takes an input string and replaces all occurrences within it of a specified string value with another string value. The function supports limited and unlimited `STRING` types. Arguments of other types are not supported.

#### Syntax

The syntax is as follows:

```sql
REPLACE ( expression, pattern, replacement )
```

`expression` specifies the string to be searched.

> `pattern` specifies the pattern to search for. In the function output, all occurrences of the pattern will be replaced with the value of `replacement`.

Reviewer: Does this have to be a specific string, or are regular expressions allowed?

Reviewer: Good change.

### Binary operators

These are the supported binary operators:

| Operator | Description              | Type       |                                                         
|----------|--------------------------|------------|
| `+`      | Addition                 | Arithmetic |
| `-`      | Subtraction              | Arithmetic |
| `*`      | Multiplication           | Arithmetic |
| `:`      | Division                 | Arithmetic |
| `%`      | Modulo                   | Arithmetic |
| `=`      | Equal to                 | Logical    |
| `!=`     | Not equal to             | Logical    |
| `<`      | Less than                | Logical    |
| `<=`     | Less than or equal to    | Logical    |
| `>`      | Greater than             | Logical    |
| `>=`     | Greater than or equal to | Logical    |
| `OR`     | Logical disjunction      | Logical    |
| `AND`    | Logical conjunction      | Logical    |

Binary operators are used with this syntax:
```sql
	expression operator expression
```
Where `operator` is any available binary operator. Both `expression` operands should be of compatible types for the operator and compatible with the other operand.

> #### Type coercion precedence

Reviewer 2: I'm referring to this using `#type-coercion` tag. Please ad it here.

Returns true is the left `expression` is less than the right. Both operands should be numeric.

It can be used for filtering data with the use of a `WHERE` clause:
```sql
SELECT LastName, Number, Price FROM Sales.Order WHERE Price < 5
```

| LastName | Number | Price |                                                         
|----------|--------|-------|
| Doe      | 7      | 1.5   |

#### <= (Less than or equal to)
Returns true is the left `expression` is less or equal to the right. Both operands should be numeric.
#### \> (Greater than)
Returns true is the left `expression` is greater than the right. Both operands should be numeric.
#### \>= (Greater than or equal to)
Returns true is the left `expression` is greater or equal to the right. Both operands should be numeric.
#### OR
Returns true if at least one input `expression` returns true. Both operands need to be of type `BOOLEAN`.
#### AND
Returns true if both input `expression` return true. Both operands need to be of type `BOOLEAN`.

Its main use is to make complex `WHERE` conditions with a combination of input values. In the query below large orders or smaller orders with a high value are selected:

```sql

> SELECT LastName, Number, Price FROM Sales.Order WHERE Number > 5 OR (Price > 5 AND Number > 2)

Reviewer 2: In an example with parentheses, I'd put `OR` in parentheses and `AND` outside to illustrate how they affect the precedence. Currently, the parentheses do not affect the result.

| `-`      | Arithmetic negation | Arithmetic |
| `NOT`    | Logical negation    | Logical    |

Unary operators are used with syntax:
```sql
	operator expression
```
`expression` should be of type compatible with the `operator`.

#### - (Arithmetic negation)
Negates a numeric value. The return type is the same as input `expression`. 

#### NOT
Reverses boolean `TRUE` values into `FALSE` and vice versa.

### Other operators {#other-operators}
These are operators that do not match the general unary or binary syntax. They are all logical operators. Those are:

| Operator | Description                                                     |     
|----------|-----------------------------------------------------------------|
| `LIKE`   | Matches a string to a specified pattern                         |
| `IN`     | Matches any value in a subquery or a list of expression values. |
| `EXISTS` | Test for the existence of any rows when executing the subquery. |
| `IS`     | Tests if a value is `NULL`                                      |

#### LIKE
Matches an `expression` to the pattern after the operator. 

##### Syntax
```sql
expression LIKE pattern
```

> Where `expression` is of type `STRING` and `pattern` is a string literal or parameter. A `NULL` pattern is treated as an empty string.

Reviewer 2: I'd mention explicitly that having a function on the right side of like does not work, but on the left side it's fine

```sql
EXISTS subquery
```
Where `subquery` is any query.

##### Examples
The operator can be used to check if an entity contains any object matching a condition.

The condition `EXISTS (SELECT * FROM Sales.Customer WHERE LastName = 'Mose')` returns `FALSE` as there are no customers with the last name `Mose`.

This query returns all customers that also have orders placed:
```sql
SELECT *
FROM Sales.Customer customer
WHERE EXISTS
    (SELECT *
    FROM Sales.Order order
    WHERE order.LastName = customer.LastName)
```

| ID | LastName | FirstName |                                                         
|----|----------|-----------|
| -  | Doe      | John      |
| -  | Moose    | Jane      |

#### IS {#is-operator}
Tests for an expression being a `NULL`. Can be inverted with an optional `NOT`. Syntax:

```sql
expression IS [ NOT ] NULL
```
Where `expression` is an expression of any datatype.

##### Examples:

`IS` operator can be used to filter out rows with values that are NULL:
```sql
	SELECT Revenue, Cost FROM Sales.Finance WHERE Revenue IS NOT NULL 
```
| Revenue | Cost |                                                       
|---------|------|
| 10      | 7    |

### CASE

> The `CASE` expression is a conditional expression, similar to if/else statements in other programming languages. Each condition is an expression that returns a Boolean result. If the condition's result is true, the value of the `CASE` expression is the result that follows the condition, and the remainder of the `CASE` expression is not processed. If the condition's result is not true, any subsequent `WHEN` clauses are examined in the same manner. If no `WHEN` condition yields true, the value of the `CASE` expression is the result of the `ELSE` clause. If the `ELSE` clause is omitted and no condition is true, the result is null.

Reviewer 2: I'd say it's more similar to `CASE` expression in other languages (the ones that support it)

Author: The simple syntax is more like switch statements, while the extended is more like if else. I can mention that

Reviewer 2: TBH, now it sounds even more complicated, so let's keep the previous version after all. It is somewhat similar to if/else after all 😅

```sql
SELECT * FROM Sales.Reports
```

| ID | Text                          |                                                         
|----|-------------------------------|
| -  | "Performance is satisfactory" |
| -  | "Order has been completed"    |

Their length can be calculated with in an extra column
```sql
SELECT Text, LENGTH(Text) as text_length FROM Sales.Reports
```

| Text                          | text_length |                                                         
|-------------------------------|-------------|
| "Performance is satisfactory" | 27          |
| "Order has been completed"    | 24          |

### LOWER{#lower-function}

#### Description

Converts all uppercase characters in a given string to lowercase.

#### Syntax

The syntax is as follows:

```sql
LOWER ( expression )
```

`expression` specifies the string to convert.

#### Example
The function is useful to enforce consistent case for all strings, especially for comparisons. The query below would return no results in case-sensitive databases, as there is only a "Doe":

```sql
SELECT * FROM Sales.Customer WHERE LastName = 'doe'
```

> Using `LOWER` this inconsistency can be fully avoided:

Reviewer 2: True, but we would also not use the index anymore. I think it's worth mentioning

| -  | 2024-05-02 00:00:00 | 2025-07-05 00:00:00 | 28      |
| -  | 2024-05-02 00:00:00 | 2024-06-02 15:12:45 | 10      |

This query uses `$range_future` to retrieve all periods that end in the future:

```sql
SELECT End, Revenue FROM Sales.Period
WHERE End > RANGEBEGIN($range_future)
```
| End                 | Revenue |
|---------------------|---------|
| 2025-07-05 00:00:00 | 28      |

### RANGEEND{#oql-rangeend}

Extracts the end value of a range parameter.

#### Syntax

```sql
RANGEEND ( $range )
```

`$range` specifies the range parameter.

#### Example
This query uses `$range` to retrieve all periods that end before the end value of `$range`:

```sql
SELECT End, Revenue FROM Sales.Period
WHERE End < RANGEEND($range)
```
| End                 | Revenue |
|---------------------|---------|
| 2024-06-02 15:12:45 | 10      |

This query uses `$range_past` to retrieve all periods that ended before the end date of `$range_past`:
```sql
SELECT End, Revenue FROM Sales.Period
WHERE End < RANGEEND($range_past)
```
| End                 | Revenue |
|---------------------|---------|
| 2024-06-02 15:12:45 | 10      |

### REPLACE

> The `REPLACE` function replaces all occurrences of a specified string value with another string value. The function supports limited and unlimited `STRING` types. Arguments of other types are not supported.

Reviewer 2: I'd mention the first parameter, something along the lines of: The `REPLACE` function takes an input string and replaces all occurrences within it of a specified string value with another string value

#### Examples

The function is useful when formatting strings in a consistent manner. A space delimited list can be converted to one with commas to be used for csv. Assume we have an entity `Sales.Raw` that contains a `STRING` field:

```sql
SELECT * FROM Sales.Raw
```

| ID | Import            |                                                         
|----|-------------------|
| -  | "6 D10 machinery" |
| -  | "1 A15 tools"     |

The text format is converted with `REPLACE`:
```sql
SELECT REPLACE(Import, ' ', ',') FROM Sales.Raw
```

| Import            |                                                         
|-------------------|
| "6,D10,machinery" |
| "1,A15,tools"     |

### ROUND

Rounds a numeric `expression` by reducing precision after the decimal point.

#### Syntax

The syntax is as follows:

```sql
ROUND ( expression , length )
```

##### expression

`expression` is any numeric expression to be rounded. If `expression` is `NULL`, the function will return `NULL`.

##### length

> `length` specifies the amount of decimals to which the `expression` must be rounded. Must be of numeric type. If the value is of type `DECIMAL`, there is database specific rounding. In PostgreSQL and MySQL, `length` is rounded. With other databases, `length` is floored. If the `length` is `NULL`, the function result will be `NULL`.

Reviewer 2: > In PostgreSQL and MySQL, `length` is rounded

I would mention explicitly that it is rounded to the closest value of the required scale. Similarly, I would explain a bit more what flooring means.

Reviewer 2: "is rounded to the nearest natural value" is not entirely correct, I  think. `ROUND(1.128, 2) == 1.13`, so it gets rounded not to a natural value, but to a value of requested scale

Author: Misunderstanding here, im writing about the value of `LENGTH` itself getting rounded if its a decimal and not an integer. I will rewrite this to be more clear.

Reviewer 2: Oh, indeed a misunderstanding. TBH, I would consider leaving this whole part out because it is a very specific detail that only makes things confusing, as we just saw.

Author: Yeah perhaps i write too many details sometimes... This can be left out

| `INTEGER`  | Integer/Long     | 5                     | Integer data                               |
| `LONG`     | Integer/Long     | 5                     | 64 bit width integer data                  |
| `STRING`   | String           | 'my_string'           | Textual data                               |

## Literals
Literals represent values that are constant and are part of the query itself. The supported literals are detailed below:

| Format | Example         | Data Type            | Description                                   |
|--------|-----------------|----------------------|-----------------------------------------------|
| -      | `TRUE`, `FALSE` | `BOOLEAN`            | Conditional constants                         |
| 's*'   | 'my_string'     | `STRING`             | String literal                                |
| d+     | 5               | `INTEGER` and `LONG` | Natural number literal                        |
| d+.d+  | 5.3             | `DECIMAL`            | Real number literal                           |
| -      | `NULL`          | -                    | NULL literal to represent non-existent values |

> Where `d` is a number, `s` is any character, * indicates zero or more characters, + indicated one or more characters.

Reviewer: Is "+" always a numeric character?

Reviewer 2: + is not a character at all. Here Pijus uses regex syntax, but apparently does not explain it clear enough.

In regex, "d+" means "one or more occurrences of d"

Reviewer: 🤦

### DATETIME
There is no direct support for `DATETIME` literals. For functions that take `DATETIME` as input, it can be represented with a `STRING` in a ISO date time format or a `LONG` value representing unix seconds.

## System variables
Xpath [system variables](/refguide/xpath-keywords-and-system-variables/) can be used in OQL with the format:
```sql
'[%SystemVariable%]'
```

These variables can be used the same way as other expressions. This query gets the names of all `Person` objects that are associated with the current user:
```sql
select Name from Person where Person_USER = '[%CurrentUser%]'
```

## Operators

> Operators perform common operations and do not use parenthesis in their syntax. They take `expression` as input, which can be other operator results, functions, columns and literals.

Reviewer: You mean you cannot force precedence using parentheses, or something else?

Reviewer: I guess you mean that they aren't functions and so parameters (`expressions`) are not in parentheses. Like it isn't `+(a, b)` but is `a+b`.
I think we could state this more clearly?
```suggestion
Operators perform common operations and, unlike functions, do not put their parameters in parentheses. They take `expression` as input, which can be other operator results, functions, columns and literals.
```

Reviewer 2: The change looks good

| `:`      | Division                 | Arithmetic |
| `%`      | Modulo                   | Arithmetic |
| `=`      | Equal to                 | Logical    |
| `!=`     | Not equal to             | Logical    |
| `<`      | Less than                | Logical    |
| `<=`     | Less than or equal to    | Logical    |
| `>`      | Greater than             | Logical    |
| `>=`     | Greater than or equal to | Logical    |
| `OR`     | Logical disjunction      | Logical    |
| `AND`    | Logical conjunction      | Logical    |

Binary operators are used with this syntax:
```sql
	expression operator expression
```
Where `operator` is any available binary operator. Both `expression` operands should be of compatible types for the operator and compatible with the other operand.

#### Type coercion precedence {#type-coercion}

Binary operations perform type casting when operand types are differing. For operations involving only numeric types, data types are always upcasted to ensure data types are matching. The result type will also be the operand type with the highest precedence. This is done according to this ordering:
- `DECIMAL`
- `LONG`

> - `INTEGER`

Reviewer: Should we refer to CAST here?

Reviewer 2: How exactly do you mean?

{{% alert color="info" %}}

This precedence rule does not apply for operations where at least one of the operands is non-numeric, including that of type `STRING`. The final result type will depend on the database.

{{% /alert %}}

#### + (Addition)
Performs different operations depending on the first `expression` datatype. A numeric input performs a arithmetic addition, while a `STRING` input performs string concatenation. 

Assume `Sales.Customer` contains two objects and `Sales.Order` contains three objects.

```sql
SELECT * FROM Sales.Customer
```

| ID | LastName | FirstName |                                                         
|----|----------|-----------|
| -  | Doe      | John      |
| -  | Moose    | Jane      |

```sql
SELECT * FROM Sales.Order
```

| ID | LastName | Number | Price |                                                         
|----|:---------|--------|-------|
| -  | Doe      | 7      | 1.5   |
| -  | Doe      | 2      | 5     |
| -  | Moose    | 3      | 8.2   |

The operator can be used to modify an attribute in SELECT.
```sql

> SELECT (Number + 5) FROM Sales.Order

Reviewer: Why does this return `LastName`?
And you say above that "Operators … do not use parentheses in their syntax" and then show an example which has parentheses around the operator expression.

| `EXISTS` | Test for the existence of any rows when executing the subquery. |
| `IS`     | Tests if a value is `NULL`                                      |

#### LIKE
Matches an `expression` to the pattern after the operator. 

##### Syntax
```sql
expression LIKE pattern
```
Where `expression` is of type `STRING` and `pattern` is a string literal or parameter. Note that this means functions are not allowed to be used in `pattern`. A `NULL` pattern is treated as an empty string.

The pattern can have special characters, which are all wildcards. Supported wildcard Characters:

| Wildcard Character | Description                           |     
|--------------------|---------------------------------------|
| `%`                | Matches zero or more of any character |
| `_`                | Matches one of any character          |

In order to search for special characters, they should be escaped with the `\` escape character (including `\` itself).

##### Examples
Presume we have 3 strings for column `House`: `Apartment`, `Tenement` and `Flat`. We can select all string ending with "ment" with this condition:
```sql

> Select House FROM House LIKE '%ment' 

Reviewer: This looks a bit ambiguous. Do the entity and attribute have the same name?

```
| House     |                                                         
|-----------|
| Apartment |
| Tenement  |

A certain length of string can be enforced with the use of the `_` operator This query matches any string that has 4 of any character ending with "mend"":
```sql
Select House FROM House LIKE '____ment' 
```
| House     |                                                         
|-----------|

> | Apartment |

Reviewer: Wouldn't this be rejected as it has five characters before "ment"?
```suggestion
```

```sql
SELECT * FROM Sales.Customer WHERE LastName = 'doe'

> ```f

Reviewer 2: Probably a typo

|----|----------|-----------|
| -  | Doe      | John      |

> Note that this query caj no longer take advantage of an index for `LastName` for comparison, resulting in a performance decrease.

Reviewer 2: `caj` is a typo
